fpga_interchange/arch_pack_clusters.cc

```diff
@@ -0,0 +1,496 @@
+        for (auto cell : bel_chain.cells) {
+            IdString cell_name(cell);
+            log_info("      - %s\n", cell_name.c_str(ctx));
+        }
+
+        log_info("  - pattern:\n");
+        IdString source_cport(bel_chain.source->cell_port);
+        IdString source_bport(bel_chain.source->bel_port);
+        IdString sink_cport(bel_chain.sink->cell_port);
+        IdString sink_bport(bel_chain.sink->bel_port);
+        log_info("      - src %s - %s --> dst %s - %s\n", source_cport.c_str(ctx), source_bport.c_str(ctx), sink_cport.c_str(ctx), sink_bport.c_str(ctx));
+
+        log_info("  - drivers:\n");
+        for (auto &driver : bel_chain.chain_drivers) {
+            log_info("    - in only cells:\n");
+            for (auto cell : driver.in_cells)
+                log_info("      - %s\n", IdString(cell).c_str(ctx));
+            log_info("    - in and out cells:\n");
+            for (auto cell : driver.in_out_cells)
+                log_info("      - %s\n", IdString(cell).c_str(ctx));
+            log_info("    - ports:\n");
+            for (auto port : driver.ports) {
+                log_info("      - %s - %s\n", IdString(port.name).c_str(ctx), IdString(port.bel).c_str(ctx));
+            }
+        }
+    }
+}
+
+void Arch::prepare_cluster(const BelChainPOD *chain)
+{
+    Context *ctx = getCtx();
+    IdString chain_name(chain->name);
+
+    // Get chainable cells
+    std::vector<CellInfo *> chainable_cells;
+    for (auto cell : sorted(ctx->cells)) {
+        CellInfo *ci = cell.second;
+        for (auto chain_cell : chain->cells) {
+            IdString chain_cell_type(chain_cell);
+            if (ci->type == chain_cell_type) {
+                chainable_cells.push_back(ci);
+            }
+        }
+    }
+
+    // Find roots from chainable cells
+    std::vector<CellInfo *> roots;
+    for (auto cell : chainable_cells) {
+        IdString sink_cell_port(chain->sink->cell_port);
+
+        PortRef driver = cell->ports[sink_cell_port].net->driver;
+
+        auto res = std::find(chainable_cells.begin(), chainable_cells.end(), driver.cell);
+        if (res == chainable_cells.end()) {
+            // We hit a root cell
+            cell->cluster.set(ctx, cell->name.str(ctx));
+            roots.push_back(cell);
+        }
+    }
+
+    // Generate unique clusters starting from each root
+    for (auto root : roots) {
+        CellInfo *next_cell = root;
+        std::vector<CellInfo *> cluster_cells;
+
+        while (next_cell != nullptr) {
+            // Find possible source type/port to follow in cluster building
+            std::pair<IdString, IdString> config;
+            IdString src_cell_port(chain->source->cell_port);
+            IdString src_bel_port(chain->source->bel_port);
+            IdString sink_cell_port(chain->sink->cell_port);
+            IdString sink_bel_port(chain->sink->bel_port);
+
+            if (next_cell == root) {
+                disconnect_port(ctx, next_cell, sink_cell_port);
```

Comment:
can you give a bit more context behind this `disconnect_port` ?

Reply:
Sure, comment added